src/validation.h

```diff
@@ -47,6 +47,8 @@ class CValidationState;
 
 struct LockPoints;
 
+static std::atomic<bool> fDIP0001ActiveAtTip{false};
```

Comment:
Well, yes and no. Yes, it doesn't guarantee 100% safety, more like 99.9999999% or so. The thing is, `fDIP0001ActiveAtTip` value is going to be changed only once - exactly at the moment of activation and chances that some function/thread is going to be accessing it at the very same millisecond are pretty tiny imo. Also, given the fact that we are reeeally far from hitting that 1MB limit, switching `fDIP0001ActiveAtTip` from `false` to `true` shouldn't have any visible consequences - it doesn't really matter if it's 1MB or 2MB limit when blocks are 10-100KB max (same for all other cases) and `DEFAULT_BLOCK_MAX_SIZE` is still 750KB (I doubt miners ever cared to tweak it yet). So the whole `fDIP0001ActiveAtTip` thing is mostly a precautious way of doing this upgrade and `atomic` is yet another "just in case" thing. I know it sounds like a compromise but imo it's a better way (given the probability) than locking `cs_main` all over the place.